app/backend/wells/utils.py

```diff
@@ -0,0 +1,151 @@
+from wells.constants import ADDRESS_COLUMNS, GEOCODER_ENDPOINT
+from thefuzz import fuzz
+
+def calculate_pid_distance_for_well(well):
+    """
+    Calculate the distance from a single well to the nearest parcel using a WFS query.
+    :param well: A well instance with latitude, longitude attributes
+    :return: Distance to the nearest parcel
+    """
+    # Base URL for the WFS request
+    base_url = "https://openmaps.gov.bc.ca/geo/pub/wfs"
+    params = {
+        "service": "WFS",
+        "version": "2.0.0",
+        "request": "GetFeature",
+        "typeName": "WHSE_CADASTRE.PMBC_PARCEL_FABRIC_POLY_SVW",
+        "outputFormat": "json",
+        "srsName": "EPSG:4326",
+        "CQL_FILTER": f"DWITHIN(geometry, POINT({well.longitude} {well.latitude}), 0.1, meters)"
+    }
+
+    # Construct the request URL
+    request_url = f"{base_url}?{'&'.join([f'{k}={v}' for k, v in params.items()])}"
+
+    # Make the request
+    response = requests.get(request_url)
+    if response.status_code != 200:
+        print("Error making request to WFS service.")
+        return None
+
+    # Load response into GeoDataFrame
+    data = response.json()
+    parcels_gdf = gpd.GeoDataFrame.from_features(data["features"])
+
+    if parcels_gdf.empty:
+        print("No parcels found near well location.")
+        return None
+
+    # Calculate the distance from the well to the nearest parcel
+    well_point = Point(well.longitude, well.latitude)
+    parcels_gdf["distance"] = parcels_gdf.distance(well_point)
+
+    return parcels_gdf["distance"].min()
+
+
+def calculate_natural_resource_region_for_well(well):
+    """
+    Retrieve the natural resource region name that a well is within using a WFS query.
+    :param well: A well instance with latitude, longitude attributes
+    :return: Natural Resource Region name
+    """
+    # Base URL for the WFS request
+    base_url = "https://openmaps.gov.bc.ca/geo/pub/wfs"
+    params = {
+        "service": "WFS",
+        "version": "2.0.0",
+        "request": "GetFeature",
+        "typeName": "WHSE_ADMIN_BOUNDARIES.ADM_NR_DISTRICTS_SPG",
+        "outputFormat": "json",
+        "srsName": "EPSG:4326",
+        "CQL_FILTER": f"CONTAINS(geometry, POINT({well.longitude} {well.latitude}))"
+    }
+
+    # Construct the request URL
+    request_url = f"{base_url}?{'&'.join([f'{k}={v}' for k, v in params.items()])}"
+
+    # Make the request
+    response = requests.get(request_url)
+    if response.status_code != 200:
+        print("Error making request to WFS service.")
+        return None
+
+    # Load response into GeoDataFrame
+    data = response.json()
+    regions_gdf = gpd.GeoDataFrame.from_features(data["features"])
+
+    if regions_gdf.empty:
+        print("No natural resource regions found near well location.")
+        return None
+
+    # Assuming the well can only be in one region, return the name of the first region found
+    return regions_gdf.iloc[0]['properties']['REGION_ORG_UNIT_NAME']
```

Comment:
Here we are checking if `regions_gdf` exists do we also need to check if `iloc` exists before indexing?

Reply:
the 'regions_gdf.empty' check above is sufficient to prevent indexing errors because it ensures that the dataframe contains data before attempting to access it.